lib/scope.js

```diff
@@ -0,0 +1,55 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+function Scope(arr) {
+  if (!(this instanceof Scope)) {
+    return new Scope(arr);
+  }
+  if (!arr) {
+    arr = [];
+  } else if (arr instanceof Scope) {
+    return arr;
+  } else if (typeof arr === 'string') {
+    arr = arr.split(' ');
+  }
+  var obj = {};
+  for (var i = 0; i < arr.length; i++) {
+    obj[arr[i]] = true;
+  }
+  this._values = obj;
+}
+
+Scope.prototype = {
+
+  _values: undefined,
+
+  has: function has(scope) {
+    var that = this;
+    return Scope(scope).values().every(function(word) {
+      if (word in that._values || word + ':write' in that._values) {
+        return true;
+      } else {
+        var prefix = word.split(':').slice(0, -1).join(':');
+        return prefix && that.has(prefix);
+      }
+    });
+  },
```

Comment:
I mean that we could allow the SP to provide the list of expected scopes when verifying the token, and check it for them by calling out to this logic, e.g.:

```
> POST /v1/verify
> {
>   "token": "blahblahblah"
>   "scopes": "profile:email readinglist:write"
> }
...
< 40X Some Error Code
< {
<   message: "Scopes did not match"
< }
```

It might reduce footguns when implementing a new Service Provider.